### DISCUSSION TITLE
Custom bootloader for the blackpillv2

### DISCUSSION THREAD

**Discussion by HrMitrev** [blackmagic-debug/blackmagic#1299] [Polls]
Hi!
I recently bought a couple of blackpill boards(STM32F411CE variant) from China. I don't know if it's because of the chip shortage or just too common problem in China, but the uC turns out to be fake. Interestingly enough I woudn't have known if I flashed blackmagic FW over the SWD port, because it so far works perfectly.
The problem is that the embedded bootloader refuses to work(with dfu-util at least). While I was searching for solution I found [this](https://github.com/qmk/qmk_firmware/issues/17576). My chips are the same fakes and I presume there are way more in the wild.

So my question is: Do you think it's a good idea to add a cutom bootloader for this board(like how it's done for the stlink target)? There are way more flash and SRAM than what it's needed. It may be an optional parameter(like how it's done with ENABLE_RTT=1)

This way there will be no problem if you got the clone with the broken embedded bootloader and upgrading the firmware will be as easy as with the stlink target when you initially flash the bootloader over SWD.

**Comment by dragonmux** [comment]
When building `make PROBE_HOST=stlink`, specifying `ST_BOOTLOADER=1` turns off the BMP bootloader which is otherwise built. With `make PROBE_HOST=swlink`, the BMP bootloader is always built.

You can load the `src/blackmagic_dfu.bin` binary onto your board in place of the factory bootloader and, while not terribly fast for now, it will give you a DFU-compliant bootloader to use

**Reply by HrMitrev** [reply]
Hi again,
I think that the bootloader I have in mind cannot be overwritten. In [AN2606](https://www.st.com/resource/en/application_note/cd00167594-stm32-microcontroller-system-memory-boot-mode-stmicroelectronics.pdf) they say "_The bootloader is stored in the internal boot ROM (system memory)_". Not sure `embedded` was the right word to describe it.
So `src/blackmagic_dfu.bin` will need to go to the address of `src/blackmagic.bin` and that binary should be moved with offset of 0x2000. I'm not too familiar with the details of the binary linking stage and the differences between STM32F411 and STM32F103. I'm not sure if the same bootloader might be reusable(it very well may be). I'm also not sure if I flash a working bootloader at 0x08000000, that the blackmagic.bin would work without change in the linker script to acount for the offset.

**Reply by dragonmux** [reply]
blackmagic.bin is designed/intended on swlink to be loaded at address 0x08002000, with the BMP bootloader at 0x08000000. Yes, those STM32 parts do have an internal bootloader but as soon as you have something in Flash (at the start of Flash, specifically), it'll jump to your code which can itself be a bootloader.

This is why you can find `LDFLAGS = $(LDFLAGS_BOOT) -Wl,-Ttext=0x8002000` in the Makefile.inc for swlink as this offsets the firmware to allow for a bootloader at the start of Flash

**Reply by HrMitrev** [reply]
Hi,
>but as soon as you have something in Flash (at the start of Flash, specifically), it'll jump to your code which can itself be a bootloader.

I know, but the blackpillv2 platform seemed not to have an option to build a bootloader for it. I looked around the make system and found out the `BMP_BOOTLOADER=1` flag. Building with `make PROBE_HOST=blackpillv2 ENABLE_RTT=1 BMP_BOOTLOADER=1` fails:
```
  BUILD   lib/stm32/f1
  BUILD   lib/stm32/f4
  BUILD   lib/lm4f
make[2]: Nothing to be done for 'all'.
make[2]: Nothing to be done for 'all'.
make[2]: Nothing to be done for 'all'.
Load address 0x08004000 for BMPBootloader
 GEN include/version.h
  CC      command.c
  CC      remote.c
  CC      platforms/common/usb.c
  CC      platforms/stm32/dfucore.c
platforms/stm32/dfucore.c: In function 'get_dev_unique_id':
platforms/stm32/dfucore.c:333:23: error: 'FLASH_BASE' undeclared (first use in this function); did you mean 'HASH_BASE'?
  333 |         max_address = FLASH_BASE + (fuse_flash_size << 10U);
      |                       ^~~~~~~~~~
      |                       HASH_BASE
platforms/stm32/dfucore.c:333:23: note: each undeclared identifier is reported only once for each function it appears in
make[1]: *** [Makefile:119: dfucore.o] Error 1
make[1]: *** Waiting for unfinished jobs....
make: *** [Makefile:17: all] Error 2
```
If I build with `make PROBE_HOST=blackpillv2 OPT_FLAGS='-DFLASH_BASE=0x8000000UL -Os' ENABLE_RTT=1 BMP_BOOTLOADER=1` it compiles, but the bootloader doesn't seem to work after I load it, my PC cannot detect any usb device.

**Reply by dragonmux** [reply]
Please report that build failure as a bug on this issue tracker as that appears to be a regression in the bootloader for that platform. You should not need to specify `FLASH_BASE` in that manner.

**Reply by HrMitrev** [reply]
I think that the blackpillv2 platform is not intended to have a bootloader at all, because you should be able to use the system bootloader and it will work exactly the same as swlink works now. I raised this discussion because of the fakes I found which will benefit from a custom bootloader. It is important to note, that the genuine chips do not need something like this.

**Reply by dragonmux** [reply]
The blackpillv2 platform was roughly split from the f4discovery platform and is intended to work identically. It is a bug that it doesn't (compiling for blackpill used to be `make PROBE_HOST=f4discovery BLACKPILL=1`)
